Vault documentation: release notes for 1.12 (#17237)
hashicorp/vault: website/content/docs/release-notes/1.12.0.mdx

Some of these enhancements and changes in this release include the following:

- Vault Enterprise now supports PKCS#11 provider plugin (client library) functionality.
- Vault Enterprise can now manage keys for Oracle TDE. This requires the Advanced Data Protection license.
- PKI Key revocation improvements are made to Vault’s PKI engine, introducing a new OCSP responder and automatic CRL rebuilding (with up-to-date Delta CRL), that offers significant performance and data transfer improvements to revocation workflows.
- BYOK in Transform engines now allow users to import their keys generated elsewhere.
- KMIP Server Profile adds support for additional operations, allowing  Vault to claim support for the baseline server profile.
- Transform secrets engine supports time-based auto-key rotation for tokenization.
- Path and Role-based Quotas extend the existing Vault Quota support by allowing quotas to be extended to the API path suffixes and auth mount roles.
- Licensing termination behavior has changed where non-evaluation licenses (production licenses) will no longer have a termination date.

> - TFVP includes improvements made to the Vault Version Detection and Auth engine.

Reviewer: Version detection is a new feature and should be "imprevments to Auth Engines".

Reviewer 2: Re-wrote the whole sentence ([commit](https://github.com/hashicorp/vault/pull/17237/commits/4583f44c3151389f3f470d6abe502b37739e04be))

## New Features

This section describes the new features introduced in Vault 1.12.0.

> ##3 Transform Secrets Engine Enhancements

Reviewer 3: ```suggestion
### Transform Secrets Engine Enhancements
```

These features need the Vault Enterprise ADP License.

> ### Bring Your Own Key(BYOK) for Transform

Reviewer 3: ```suggestion
### Bring Your Own Key (BYOK) for Transform
```

In release 1.11, we introduced BYOK support to Vault, enabling  customers to import existing keys into the Vault Transit Secrets Engine and  enabling secure and flexible Vault deployments.
We  are extending that support to the Vault Transform Secrets Engine in this release.

### MSSQL Support

An MSSQL store is now available to be used as an external storage engine with tokenization Transform Secrets Engine. Refer to the following documents, [Transform Secrets Engine(API)](/api-docs/secret/transform), [Transform Secrets Engine](/docs/secrets/transform/index), and [Tokenization Transform](/docs/secrets/transform/tokenization) for more information.

### Key Auto Rotation

Periodic rotation of encryption keys is a recommended key management practice for a good security posture. In Vault release 1.10, we added support for Auto key rotation for Transit Secrets Engine. In Vault 1.12, the Transform secrets engine is now  enhanced, allowing users to set the rotation policy during key creation in a time interval, which will cause Vault to rotate the Transform keys when the time interval elapses automatically.

> Refer to the following documents, [Transit Secrets Engine](/docs/secrets/transit) and [Transit Secrets Engine(API)](/api-docs/secret/transit#rotate-key) for more information.

Reviewer 3: @schultz-is / @rculpepper -- should this point to Transform rotate-key docs perhaps?

Reviewer 4: ah yes, good catch. key auto-rotation functionality was added for transform in this release

Author: @schultz-is , hi. Please provide the doc link and I'll replace the existing link(s). Thanks.

Reviewer 4: ```suggestion
Refer to the following documents, [Tokenization Transform](/docs/secrets/transform/tokenization) and [Transform Secrets Engine(API)](/api-docs/secret/transform#rotate-tokenization-key) for more information.
```

@taoism4504 sure thing! these will probably cover it

In this release, the existing resource quota functionality has been enhanced. In addition to applying the API rate limiting and lease quotas at the namespace or mount level, you can now use  the quotas to the [API path suffixes and auth mount roles](/docs/enterprise/lease-count-quotas). This enhancement provides users with more control over issued certificates.

### Client Count Improvements

The billing period for client counting API can now be specified with the [current month](/docs/concepts/client-count) for the end date parameter. When this is done the "new_clients" field will have an hyperlog approximate value indicating the number of new clients that came in the current month. Note that for the previous months, the number will be an exact value.

### Redis Secrets Engine

With the support of the Redis Secrets Engine, users can use Vault to manage static or dynamic credentials for Redis OSS. The engine  works similarly to other database secrets engines. Refer to the [Redis](/docs/secrets/redis) documentation for more information. Huge thanks to [Francis Hitchens](https://github.com/fhitchen), who contributed their repository to HashiCorp

### AWS Elasticache Secrets Engine

> With the support of the AWS ElastiCache Secrets Engine, users may use Vault to manage static  for AWS Elasticache instances that use Redis as their backend. The engine will work similarly to other database secrets engines. Refer to the [elasticache](/docs/secrets/databases/rediselasticache) documentation for more information.

Reviewer: static roles...missing the role part.

Reviewer 2: Fixed ([commit](https://github.com/hashicorp/vault/pull/17237/commits/4583f44c3151389f3f470d6abe502b37739e04be))

Some of these enhancements and changes in this release include the following:

- Vault Enterprise now supports **PKCS#11** provider plugin (client library) functionality.
- Vault Enterprise can now manage keys for **Oracle TDE**. This requires the Advanced Data Protection license.
- **PKI Key revocation** improvements are made to Vault’s PKI engine, introducing a new OCSP responder and automatic CRL rebuilding (with up-to-date Delta CRL), that offers significant performance and data transfer improvements to revocation workflows.
- **BYOK in Transform engines** now allow users to import their keys generated elsewhere.
- **KMIP Server Profile** adds support for additional operations, allowing  Vault to claim support for the baseline server profile.
- **Transform secrets engine** supports time-based auto-key rotation for tokenization.
- **Path and Role-based Quotas** extend the existing Vault Quota support by allowing quotas to be extended to the API path suffixes and auth mount roles.
- **Licensing** termination behavior has changed where non-evaluation licenses (production licenses) will no longer have a termination date.

> - **Terraform Vault Provider** improvements include Version Detection. It allows Terraform Vault Provider to detect the target Vault version to determine if the selected Vault feature is available.

Reviewer 5: Is it generally okay to mention changes that are not directly in the Vault binary release (such as this one, which is on TFVP)?

Reviewer 2: ☝️ @narayan-iyengar Any thought on this?

Reviewer: I didn't catch this. I think the genral guidance is Vault release notes only contain Vault changes. Release blogs contain features NOT in Vault binary.

Reviewer 5: Sounds good. Let's swap this with the new Database Secret Engines that we're adding then :)

> 

Reviewer 5: Should we mention Redis and Elasticache Database Secrets Engine here since they are net new engines/integrations that we will be supporting?

Reviewer: yeah we should mention those engines

### SSH Secrets Engine Support for Generating Keys

Previously, Vault's SSH Secrets Engine when used as an SSH CA required requesters to provide their own public key for signing. In Vault 1.12, Vault can now generate credential key pairs dynamically, returning them to the requester.

This was a community contributed enhancement.

### Path and Role-Based Resource Quotas

In this release, the existing resource quota functionality has been enhanced. In addition to applying the API rate limiting and lease quotas at the namespace or mount level, you can now use  the quotas to the [API path suffixes and auth mount roles](/docs/enterprise/lease-count-quotas). This enhancement provides users with more control over issued certificates.

### Client Count Improvements

The billing period for client counting API can now be specified with the [current month](/docs/concepts/client-count) for the end date parameter. When this is done the "new_clients" field will have an hyperlog approximate value indicating the number of new clients that came in the current month. Note that for the previous months, the number will be an exact value.

> ### Redis Secrets Engine

Reviewer 5: ```suggestion
### Redis Database Secrets Engine
```

> With the support of the Redis secrets engine, users can use Vault to manage static role or dynamic credentials for Redis OSS. The engine  works similarly to other database secrets engines. Refer to the [Redis](/docs/secrets/redis) documentation for more information. Huge thanks to [Francis Hitchens](https://github.com/fhitchen), who contributed their repository to HashiCorp

Reviewer 5: ```suggestion
With the support of the Redis database secrets engine, users can use Vault to manage static role or dynamic credentials for Redis OSS. The engine  works similarly to other database secrets engines. Refer to the [Redis](/docs/secrets/redis) documentation for more information. Huge thanks to [Francis Hitchens](https://github.com/fhitchen), who contributed their repository to HashiCorp
```

> ### AWS Elasticache Secrets Engine

Reviewer 5: ```suggestion
### AWS Elasticache Database Secrets Engine
```

> With the support of the AWS ElastiCache secrets engine, users may use Vault to manage static  for AWS Elasticache instances that use Redis as their backend. The engine will work similarly to other database secrets engines. Refer to the [elasticache](/docs/secrets/databases/rediselasticache) documentation for more information.

Reviewer 5: ```suggestion
With the support of the AWS ElastiCache database secrets engine, users may use Vault to manage static  for AWS Elasticache instances that use Redis as their backend. The engine will work similarly to other database secrets engines. Refer to the [elasticache](/docs/secrets/databases/rediselasticache) documentation for more information.
```

### LDAP Secrets Engine

> Vault 1.12 introduces a new LDAP secrets engine that unifies the user experience between the Active Directory (AD) secrets engine and OpenLDAP secrets engine. This new engine simplifies the user experience when Vault is used to manage credentials for directory services. This new engine supports all modes from both of the engines mentioned above (AD, LDAP and RACF modes) and brings dynamic credential capabilities for users relying on Active Directory.

Reviewer 5: ```suggestion
Vault 1.12 introduces a new LDAP secrets engine that unifies the user experience between the Active Directory (AD) secrets engine and OpenLDAP secrets engine. This new engine simplifies the user experience when Vault is used to manage credentials for directory services. This new engine supports all implementations from both of the engines mentioned above (AD, LDAP and RACF) and brings dynamic credential capabilities for users relying on Active Directory.
```
